### ISSUE TITLE
First `goto()` call after hot reload and page refresh causes destroy-and-mount and state loss

### ISSUE THREAD

**Issue by rjkip** [sveltejs/kit#1579]
**Describe the bug**
First `goto()` call after hot reload and page refresh causes a destroy-and-mount and state loss. This state loss and destroy-and-mount can be annoying when developing and the page has heavy initialisation, like network calls and a slippy map in my case.

**Logs**

–

**To Reproduce**

1. Clone https://github.com/rjkip/sveltekit-destroy-and-mount.
2. Run `npm run dev`.
3. Open `http://localhost:3000`; it redirects to `/0`. `🧛 Count Mount!` is logged to the console.
4. Click the button a few times; nothing is logged to the console.
5. Cause the page to hot-reload by eg. uncommenting the second `++count;`.
6. Refresh the page.
7. Click the button.
8. The page's state is lost and `🧛 Count Destroy!` and `🧛 Count Mount!` are logged to the console.

You may to retry a handful of times and play with hot-reloading/refreshing. I reproduced this several times when writing up this bug, but when I tried again before clicking submit the issue wasn't there...

**Expected behavior**

8. The button always increments the count on first click without losing state and resetting to 0, and the component stays mounted.

**Stacktraces**

–

**Information about your SvelteKit Installation:**

<details>
  <summary>Diagnostics</summary>

- The output of `npx envinfo --system --npmPackages svelte,@sveltejs/kit,vite --binaries --browsers`

      System:
        OS: Linux 4.19 Ubuntu 20.04.2 LTS (Focal Fossa)
        CPU: (8) x64 Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz
        Memory: 364.20 MB / 12.39 GB
        Container: Yes
        Shell: 5.8 - /usr/bin/zsh
      Binaries:
        Node: 12.22.1 - /usr/bin/node
        npm: 7.10.0 - ~/.npm-prefix/bin/npm
      Browsers:
        Firefox: 88.0.1
      npmPackages:
        @sveltejs/kit: next => 1.0.0-next.109
        svelte: ^3.34.0 => 3.38.2

- Your browser

  Firefox 88.0 (64-bit)

- Your adapter (e.g. Node, static, Vercel, Begin, etc...)

  Node
</details>

**Severity**
Just annoying, really.

**Additional context**

–

**Comment by efreibe** [comment]
I have the seme problem, I made a simple example with 3 pages and two buttons.

Initially I navigate to the url `http://localhost:3000/site/recibos`, the page is loaded I can see my __layout is loaded in the network tab:
![image](https://user-images.githubusercontent.com/1061126/126258346-230d4343-3631-4f0f-b4e0-848ca5093d40.png)

Then I click the button that executes the following command:
```html
<button on:click="{() => goto('/site/empresas')}">Empresas</button>
```
then I see in the network tab that the __layout is loaded again:
![image](https://user-images.githubusercontent.com/1061126/126258484-4ae47dda-a6f3-4413-99ab-5821630d13af.png)

After that subsequent links are working as expected:
```html
<button on:click="{() => goto('/site/legajos')}">Legajos</button>
```
![image](https://user-images.githubusercontent.com/1061126/126258655-69c897d5-7e8e-43c5-9cda-a76fe75bd27a.png)

**Comment by ignatiusmb** [comment]
I can't reproduce even with `goto` commented out, can you try and see if this happens with the default vite template?

**Comment by neckaros** [comment]
i'm experiencing the same behavior than @efreibe first goto() trigger a 'page refresh', subsequent are okay

**Comment by jdin** [comment]
I have created a minimal repo with the instructions to reproduce the bug: [https://github.com/jdin/svelte-kit-reset-bug](https://github.com/jdin/svelte-kit-reset-bug)

I hope this will help to fix it quickly. 😊

**Comment by kazzkiq** [comment]
This is not only happening with `goto()`, but with the first click on an `<a>` element also.

Only happens a single time (first click/`goto`), after that everything works as expected.

**Comment by petebot** [comment]
I'm also experiencing this issue on `<a>` elements as mentioned by @kazzkiq.

**Comment by mihrich** [comment]
Guys, I have the same problem. Lost state for the first time

**Comment by kazzkiq** [comment]
Good news: Just noticed it only happens on dev server.

I've deployed my application with both `adapter-static` and `adapter-node` and after build it works as expected (no reloads on first `goto()` or `<a>` click).

So it seems like this bug happens only on development, no issues with production builds.

**Comment by mihrich** [comment]
> Good news: Just noticed it only happens on dev server.
> 
> I've deployed my application with both `adapter-static` and `adapter-node` and after build it works as expected (no reloads on first `goto()` or `<a>` click).
> 
> So it seems like this bug happens only on development, no issues with production builds.

I don’t think so!
I have version **SvelteKit v1.0.0-next.338** and I still have the state reset for the `npm run dev` and `npm run preview` commands
Do pages with the same content count as different pages?
For example, there are the following routes: **page** and **page/id**
If I go from page **/page** to **/page/id** using `goto(/page/${id})` then the first time my state is reset.
Subsequent transitions work without resetting the state.

**Comment by neckaros** [comment]
@Rich-Harris i saw you put the need clarification label. How can we help?

**Comment by bosticka** [comment]
After upgrading my app to the new routing and the lastest sveltkit 1.0.0-next.524, I am seeing this issue. 

It only occurs after the node server restarts and then on the first goto, you get a full page refresh.  After that, you can load the 
site on new tabs etc and it works fine until you restart the node server again.

Definitely an initialization issue, maybe Vite related etc...  Anyone have a workaround?

**Comment by bosticka** [comment]
More testing and I realized that this only happens when I goto a routeId that is importing a new component. 

If I called goto to a routeId that was simple html, no problem.  But importing that first component for the first time was causing the reload.

My workaround is to pre cache my main route components before I call goto by just importing them in the main +layout.svelte.

**Comment by Rich-Harris** [comment]
This isn't a bug, it's just how HMR works. If you edit a component, any instances of that component (and their children) will be destroyed and replaced with the new one. State preservation only applies to sibling/parent components — attempting to preserve state within a component that you're editing is a fool's errand, and leads to far more confusing/unwanted behaviour than remounting.

**Comment by FeldrinH** [comment]
I just checked and I can reproduce the original issue using a pre 1.0 version of SvelteKit, however, after updating to SvelteKit 1.10.0 (latest at the time of writing), I cannot seem to reproduce the issue. It is possible that the issue has been fixed in one of the newer versions of SvelteKit. As pointed out in the original post, there seems to be some element of luck involved in getting the issue to occur, so I wouldn't say with certainty that the issue is resolved.
